### DISCUSSION TITLE
Reapplying Free Highlights

### DISCUSSION THREAD

**Discussion by masyron** [stephanrauh/ngx-extended-pdf-viewer#2663] [Q&A]
Hey there! I've been working with the ngx-extended-pdf-viewer in an attempt to have the ability to reapply highlights that have been previously exported. For text highlighting, I've been able to get it working pretty smoothly. My issue comes with the Free Highlights, when I try to reapply them, it seems like it's missing the context of the box it that it was originally in if that makes sense.

So I guess my question, what piece am I missing or where should I focus looking to get that box to get free highlight positioned correctly on the page?

Thanks for your time.

**Comment by stephanrauh** [comment]
Well, the disappointing answer is that the base library doesn't implement this feature. My annotations API highjacks the copy-and-paste API. According to the pdf.js team, it doesn't make sense to copy and paste a highlight. Come to think of it, it makes sense. Therefore, they didn't implement pasting highlights from the clipboard - and my algorithm fails.

Also see #2345.

**Reply by stephanrauh** [reply]
BTW, I'm surprised that you've managed to reapply text highlights. I thought this is impossible. How did you managed to do it?

**Reply by masyron** [reply]
I was working with the highlight.js of pdf.js and had to modify the deserialization of the serialized annotation. The solution may be specific to how I'm using it. I'm also able use bounding boxes generated by the OCR process and display custom highlights on the PDF. I had to adjust some of the math to calculate the points within the PDF a little. But I was able to get it working.

Yes, that's what I kept seeing. I was hoping maybe a little direction from someone who knows more of pdf.js as to what would need to be included in the exporting of the free highlight to be able to reapply and I make the modifications myself. I tried digging through pdf.js a bit and I see where the free highlights are getting deserialized, it's just that box or container that the highlight needs to exist, I can't seem to get it to recalculate correctly. 

Any suggestions or may it just be easier to ask directly in the pdf.js team?

**Reply by stephanrauh** [reply]
@Calixteman Maybe this discussion is interesting to you. Would you like to chime in?
